File: pattern-diagnose/assets/examples/go/factory-method/before/export_service.go
// SPDX-License-Identifier: MIT
// Before: conditional construction — Factory Method candidate

package factorymethod

import (
	"encoding/json"
	"fmt"
)

// Document is the data to export.
type Document struct {
	Title   string
	Content string
}

// ExportService chooses exporter via if/else — each new format modifies this method.
type ExportService struct{}

func (s *ExportService) Export(doc Document, format string) (string, error) {
	// Conditional construction: each new format branch touches this code
	if format == "json" {
		b, _ := json.Marshal(map[string]string{"title": doc.Title, "body": doc.Content})
		return string(b), nil
	} else if format == "csv" {
		return fmt.Sprintf(`"title","content"%s"%s","%s"`, "\n", doc.Title, doc.Content), nil
	} else if format == "xml" {
		return fmt.Sprintf("<doc><title>%s</title><body>%s</body></doc>", doc.Title, doc.Content), nil
	}
	return "", fmt.Errorf("unknown format: %s", format)
}

File: pattern-diagnose/assets/examples/python/factory-method/before/export_service.py
# SPDX-License-Identifier: MIT
"""Before: conditional construction — Factory Method candidate."""

from dataclasses import dataclass


@dataclass
class Document:
    title: str
    content: str


class ExportService:
    def export(self, doc: Document, format: str) -> str:
        # Conditional construction: each new format branch touches this code
        if format == "json":
            import json
            return json.dumps({"title": doc.title, "body": doc.content})
        elif format == "csv":
            return f'"title","content"\n"{doc.title}","{doc.content}"'
        elif format == "xml":
            return f"<doc><title>{doc.title}</title><body>{doc.content}</body></doc>"
        else:
            raise ValueError(f"Unknown format: {format}")
